File: app/src/main/java/com/sleticalboy/dailywork/util/ListUtils.java
package com.sleticalboy.dailywork.util;

import java.util.List;

/**
 * Created on 18-2-7.
 *
 * @author sleticalboy
 * @version 1.0
 * @description
 */
public class ListUtils {

    public static <E> void exchangeItem(List<E> source, int from, int to) {
        if (isEmpty(source)) {
            return;
        }
        E fromItem = source.get(from);
        E toItem = source.get(to);
        source.set(from, toItem);
        source.set(to, fromItem);
    }

    public static <E> boolean equals(List<E> actual, List<E> excepted) {
        if (actual == null) {
            return excepted == null;
        }
        if (excepted == null) {
            return false;
        }
        if (actual.size() != excepted.size()) {
            return false;
        }
        for (int i = 0; i < actual.size(); i++) {
            if (!ObjectUtils.equals(actual.get(i), excepted.get(i))) {
                return false;
            }
        }
        return true;
    }

    public static <E> int getSize(List<E> source) {
        return isEmpty(source) ? 0 : source.size();
    }

    public static <E> boolean isEmpty(List<E> source) {
        return source == null || source.size() == 0;
    }

    public static <E> void relocation(List<E> source, int from, int to) {
        if (isEmpty(source)) {
            return;
        }
        E fromItem = source.remove(from);
        source.add(to, fromItem);
    }
}

File: MainApp/src/main/java/com/sleticalboy/util/ListUtils.kt
package com.sleticalboy.util

/**
 * Created on 18-2-7.
 *
 * @author leebin
 * @version 1.0
 * @description
 */
object ListUtils {

    fun <E> exchangeItem(source: MutableList<E>, from: Int, to: Int) {
        if (isEmpty(source)) {
            return
        }
        val fromItem = source[from]
        val toItem = source[to]
        source[from] = toItem
        source[to] = fromItem
    }

    fun <E> equals(actual: List<E>?, excepted: List<E>?): Boolean {
        if (actual == null) {
            return excepted == null
        }
        if (excepted == null) {
            return false
        }
        if (actual.size != excepted.size) {
            return false
        }
        for (i in actual.indices) {
            if (!ObjectUtils.equals(actual[i], excepted[i])) {
                return false
            }
        }
        return true
    }

    fun <E> getSize(source: List<E>): Int {
        return if (isEmpty(source)) 0 else source.size
    }

    fun <E> isEmpty(source: List<E>?): Boolean {
        return source == null || source.isEmpty()
    }

    fun <E> relocation(source: MutableList<E>, from: Int, to: Int) {
        if (isEmpty(source)) {
            return
        }
        val fromItem = source.removeAt(from)
        source.add(to, fromItem)
    }
}
